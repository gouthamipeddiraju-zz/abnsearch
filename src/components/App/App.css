html {
  margin: 0;
  padding: 0;
  height: 100%;
}
body {
  font-family: sans-serif;
  margin: 0;
  padding: 0;
  height: 100%;
}

#root {
  height: 100%;
}

.wrapper {
  /* height: 100%; */
  background: white;
}

.topSection {
  text-align: center;
}

input {
  border: none;
  padding: 0.5rem;
  width: 40%;
  border-bottom: 2px solid #333;
  font-size: 18px;
  outline: none;
}
input:focus {
  border-bottom: 4px solid #333;
}

.inputSection {
  text-align: center;
}

.suggestionsSection {
  width: 40%;
  text-align: center;
  margin: 0 auto;
}

.suggestions {
  border-top-width: 0;
  list-style: none;
  margin-top: 0;
  overflow-y: auto;
  padding-left: 0;
  width: 100%;
  box-shadow: 0px 2px 4px -1px rgba(0, 0, 0, 0.2),
    0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12);
}

.suggestions li {
  padding: 0.5rem;
}

.suggestionActive,
.suggestions li:hover {
  background-color: #1776d2;
  color: #ffffff;
  cursor: pointer;
  font-weight: 700;
}

.suggestions li:not(:last-of-type) {
  border-bottom: 1px solid #999;
}

.noSuggestions {
  color: #999;
  padding: 0.5rem;
}

.resultSection {
  margin-top: 20px;
  text-align: center;
  margin: 0 auto;
  min-width: 75%;
  padding: 20px;
}

table {
  margin: 0;
  width: 100%;
  text-align: center;
  box-shadow: 0px 2px 4px -1px rgba(0, 0, 0, 0.2),
    0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12);
}

tr {
  border: 3px solid black;
}

td {
  padding: 10px;
  border-bottom: 1pt solid #0000001a;
}
