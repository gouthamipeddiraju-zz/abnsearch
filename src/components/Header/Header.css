.headerWrapper {
  background-color: #1976d2;
  height: 70px;
  text-align: center;
  border: 1px solid #999;
  box-shadow: 0px 2px 4px -1px rgba(0, 0, 0, 0.2),
    0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12);
}

.headerText {
  color: #fff;
  font-size: 23px;
  padding: 20px;
}
