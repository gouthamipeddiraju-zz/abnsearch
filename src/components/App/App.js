import React, { Component, Fragment } from "react";
import Header from "../Header/Header";
import "./App.css";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeSuggestion: 0,
      filteredSuggestions: [],
      showSuggestions: false,
      userInput: "",
      loading: false,
      guid: "b6242120-5bce-4b10-9839-d3045a7682da",
      disabled: false,
      selectedEntity: {}
    };
  }

  onChange = e => {
    const userInput = e.currentTarget.value;
    if (userInput && userInput.length > 2) {
      const { guid } = this.state;
      let fetchAbnDetails = false;
      let filteredSuggestions = [];
      let regex = /^[0-9]*$/gm;
      if (userInput.match(regex)) {
        fetchAbnDetails = true;
      }
      let fetchUrl =
        "https://abr.business.gov.au/json/MatchingNames.aspx?name=" +
        userInput +
        "&maxResults=10&callback=callback&guid=" +
        guid;

      if (fetchAbnDetails) {
        fetchUrl =
          "https://abr.business.gov.au/json/AbnDetails.aspx?abn=" +
          userInput +
          "&callback=callback&guid=" +
          guid;
      }

      this.setState({
        loading: true,
        userInput,
        disabled: true,
        selectedEntity: {}
      });
      return fetch(fetchUrl)
        .then(resp => resp.text())
        .then(
          result => {
            try {
              result = result.replace("callback(", "").replace(/\)/g, "");
              result = JSON.parse(result);
              if (!fetchAbnDetails) {
                filteredSuggestions = result["Names"].filter(suggestion => {
                  return (
                    suggestion["Name"]
                      .toLowerCase()
                      .indexOf(userInput.toLowerCase()) > -1
                  );
                });
              } else {
                result && result.Abn && filteredSuggestions.push(result);
              }
              this.setState({
                activeSuggestion: 0,
                filteredSuggestions,
                showSuggestions: true,
                userInput,
                loading: false,
                disabled: false,
                selectedEntity: {}
              });
              this.nameInput.focus();
            } catch (e) {
              console.log(e);
              this.setState({
                activeSuggestion: 0,
                filteredSuggestions: [],
                showSuggestions: true,
                userInput,
                loading: false,
                disabled: false,
                selectedEntity: {}
              });
              this.nameInput.focus();
            }
          },
          error => {
            console.log(error);
            this.setState({
              activeSuggestion: 0,
              filteredSuggestions: [],
              showSuggestions: true,
              userInput,
              loading: false,
              disabled: false,
              selectedEntity: {}
            });
            this.nameInput.focus();
          }
        );
    } else {
      this.setState({
        userInput,
        disabled: false,
        selectedEntity: {}
      });
    }
  };

  onClick = suggestion => {
    this.setState({
      activeSuggestion: 0,
      filteredSuggestions: [],
      showSuggestions: false,
      userInput: suggestion["Name"]
        ? suggestion["Name"]
        : suggestion["EntityName"],
      loading: false,
      selectedEntity: suggestion
    });
  };

  onKeyDown = e => {
    const { activeSuggestion, filteredSuggestions } = this.state;

    if (e.keyCode === 13) {
      this.setState({
        activeSuggestion: 0,
        showSuggestions: false,
        userInput: filteredSuggestions[activeSuggestion].Name,
        loading: false
      });
      this.onClick(filteredSuggestions[activeSuggestion]);
    } else if (e.keyCode === 38) {
      if (activeSuggestion === 0) {
        return;
      }

      this.setState({ activeSuggestion: activeSuggestion - 1 });
    } else if (e.keyCode === 40) {
      if (activeSuggestion - 1 === filteredSuggestions.length) {
        return;
      }

      this.setState({ activeSuggestion: activeSuggestion + 1 });
    }
  };

  render() {
    const {
      onChange,
      onClick,
      onKeyDown,
      state: {
        activeSuggestion,
        filteredSuggestions,
        showSuggestions,
        userInput,
        loading,
        disabled,
        selectedEntity
      }
    } = this;
    let suggestionsListComponent;

    if (showSuggestions && userInput) {
      if (filteredSuggestions.length) {
        suggestionsListComponent = (
          <ul className="suggestions">
            {filteredSuggestions.map((suggestion, index) => {
              let className;

              // Flag the active suggestion with a class
              if (index === activeSuggestion) {
                className = "suggestionActive";
              }

              return (
                <li
                  className={className}
                  key={index}
                  onClick={() => onClick(suggestion)}
                >
                  {suggestion["Name"]
                    ? suggestion["Name"]
                    : suggestion["EntityName"]}
                </li>
              );
            })}
          </ul>
        );
      } else {
        suggestionsListComponent = (
          <div className="noSuggestions">
            <em>No suggestions found.</em>
          </div>
        );
      }
    }

    return (
      <Fragment>
        <Header />
        <div className="wrapper">
          <div className="topSection">
            <h1 className="mainHeading">Search for an Entity below</h1>
          </div>
          <section>
            <div className="inputSection">
              <input
                ref={input => {
                  this.nameInput = input;
                }}
                type="text"
                onChange={onChange}
                onKeyDown={onKeyDown}
                value={userInput}
                disabled={disabled ? "disabled" : ""}
                placeholder="Enter either ABN or company name..."
              />
            </div>
            <div className="suggestionsSection">
              {loading ? <h4>Fetching....</h4> : suggestionsListComponent}
            </div>
          </section>
          <div className="resultSection">
            {Object.keys(selectedEntity).length > 0 && (
              <table>
                <tbody>
                  {Object.keys(selectedEntity).map(key => {
                    return (
                      <tr key={key}>
                        <td>{key}</td>
                        <td>{selectedEntity[key]}</td>
                      </tr>
                    );
                  })}
                </tbody>
              </table>
            )}
          </div>
        </div>
      </Fragment>
    );
  }
}

export default App;
