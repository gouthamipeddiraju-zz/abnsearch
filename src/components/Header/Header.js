import React from "react";
import './Header.css';

class Header extends React.Component {
    render() {
      return (
        <div className="headerWrapper">
          <div className="headerText">ABN Search tool</div>
        </div>
      );
    }
}

export default Header;