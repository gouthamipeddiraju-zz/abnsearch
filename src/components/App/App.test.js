import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App from './App';
import Header from "../Header/Header";

describe('App component', () => {
    beforeEach(function() {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({Abn: "38353028748",
      AbnStatus: "0000000002",
      IsCurrent: true,
      Name: "COL CO",
      NameType: "Trading Name",
      Postcode: "0832",
      Score: 100,
      State: "NT"}));
    });
    it('Test rendering', () => {
      const component = renderer.create(<App />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders the inner component', () => {
      const wrapper = shallow(<App />);
      expect(wrapper.find(Header).length).toEqual(1);
    });
    it('renders the input', () => {
      const wrapper = shallow(<App />);
      expect(wrapper.find('input').length).toEqual(1);
    });
    it('check heading', () => {
      const wrapper = shallow(<App />);
      const counterWrapper = wrapper.find('h1');
      expect(counterWrapper.text()).toEqual('Search for an Entity below');
    });
    it('check input change', () => {
      const wrapper = shallow(<App />);
      wrapper.find('input').simulate('change', { currentTarget: { value: 'test' }, value:'test' })
      wrapper.instance().onChange({ currentTarget: { value: 'test' }, value:'test' }).then(response => {
        expect(response.NameType).toBe("Trading Name");  
      });
      expect(wrapper).toMatchSnapshot();    
    });
    it('check input onClick', () => {
      const wrapper = shallow(<App />);
      wrapper.find('input').simulate('click', {})
      wrapper.instance().onClick({});
      expect(wrapper).toMatchSnapshot();    
    });
    it('check input onKeyDown', () => {
      const wrapper = shallow(<App />);
      wrapper.find('input').simulate('change', {currentTarget: { value: 'test' }, value:'test' })
      wrapper.instance().onKeyDown({});
      expect(wrapper).toMatchSnapshot();    
    });
});