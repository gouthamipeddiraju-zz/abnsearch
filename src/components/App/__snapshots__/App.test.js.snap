// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Test rendering 1`] = `
Array [
  <div
    className="headerWrapper"
  >
    <div
      className="headerText"
    >
      ABN Search tool
    </div>
  </div>,
  <div
    className="wrapper"
  >
    <div
      className="topSection"
    >
      <h1
        className="mainHeading"
      >
        Search for an Entity below
      </h1>
    </div>
    <section>
      <div
        className="inputSection"
      >
        <input
          disabled=""
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Enter either ABN or company name..."
          type="text"
          value=""
        />
      </div>
      <div
        className="suggestionsSection"
      />
    </section>
    <div
      className="resultSection"
    />
  </div>,
]
`;

exports[`App component check input change 1`] = `ShallowWrapper {}`;

exports[`App component check input onClick 1`] = `ShallowWrapper {}`;

exports[`App component check input onKeyDown 1`] = `ShallowWrapper {}`;
